engineapi/proxy.go

```diff
@@ -245,46 +689,45 @@ func (p *Proxy) Run() {
 	}
 }
 
-func (p *Proxy) sync() bool {
+func (p *Proxy) sync() error {
 	im, err := p.ds.GetInstanceManagerRO(p.InstanceManagerName)
 	if err != nil {
 		if datastore.ErrorIsNotFound(err) {
-			p.logger.Info("Stopping proxy because the instance manager no longer exists")
-			return true
+			return errors.Errorf("instance manager no longer exists")
 		}
 
-		utilruntime.HandleError(errors.Wrapf(err, "fail to get instance manager %v for proxy sync", p.InstanceManagerName))
-		return false
+		utilruntime.HandleError(errors.Wrapf(err, "fail to get instance manager %v for engine client proxy sync", p.InstanceManagerName))
+		return nil
 	}
 
 	hasNewOwnerID := im.Status.OwnerID != p.InstanceManagerControllerID
 	if hasNewOwnerID {
-		p.logger.Infof("Stopping proxy because the instance manager changed ownership from %v to %v", p.InstanceManagerControllerID, im.Status.OwnerID)
-		return true
+		return errors.Errorf("instance manager changed ownership from %v to %v", p.InstanceManagerControllerID, im.Status.OwnerID)
 	}
 
-	return false
+	return nil
 }
 
 func (p *Proxy) DirectToURL(e *longhorn.Engine) string {
 	if e == nil {
-		p.logger.Debug("BUG: cannot get proxy direct URL with nil engine object")
+		p.getLogger().Debug("BUG: cannot get engine client proxy re-direct URL with nil engine object")
 		return ""
 	}
 
 	return imutil.GetURL(e.Status.StorageIP, e.Status.Port)
 }
 
-func (p *Proxy) Ping() (err error) {
-	return p.grpcClient.Ping()
-}
-
 func (p *Proxy) Name() string {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
```

Comment:
No need to lock here, also `InstanceManagerName` already publicly exported, so do we still need a Name method? or you can just make `instanceManagerName` private.